### DISCUSSION TITLE
MNI Registration changes real volumes

### DISCUSSION THREAD

**Discussion by qwang0225** [ANTsX/ANTs#1645] [Q&A] [answered]
Hi, I have been exploring MRI registration recently. After registering a Nifti format MRI image on MNI152 template, I found out that real total volumes of the image increased, which is not supposed to happen to my understanding. The real volume = total number of voxels * real volume of each voxel. Please see the code below.

```
fixed_image = ants.image_read(template_path)
moving_image = ants.image_read(data_path)
transform = ants.registration(fixed_image,moving_image,'Affine')

reg3t = ants.apply_transforms(fixed_image,moving_image,transform['fwdtransforms'][0])
ants.image_write(reg3t, output_path)

mov_nii = nib.load(data_path)
out_nii = nib.load(output_path)

print(mov_nii.shape, out_nii.shape)
print(mov_nii.header.get_zooms(), out_nii.header.get_zooms())
print(np.prod(mov_nii.header.get_zooms()) * np.prod(mov_nii.shape), np.prod(out_nii.header.get_zooms()) * np.prod(out_nii.shape))
```
Here is the output.
```

(176, 198, 160)        (181, 217, 181)
(0.937, 0.937, 0.937)     (1.0, 1.0, 1.0)
4586871.7        7109137.0
```

So clearly registration changed the image spacing, so the shape of image and unit voxel changed, but I would expect the total volume remain the same. The reason I'm asking is that I try to calculate a segmented volume after feeding the registered image into a binary segmentation model. I would basically multiple the number of positive voxels and the volume of unit pixel based on output mask. 
I want to make sure that this way I will get the real volume of the segmentation.

Can any of you help out? Where did I miss? Thanks.

**Comment by cookpa** [comment]
An affine transform does not preserve volumes. It can scale the moving image to match the fixed image.

A rigid transform will preserve volumes, however there will be some imprecision with discretely sampled labels after resampling into the fixed space, so they won't be identical but should be closer.

**Reply by qwang0225** [reply]
Thanks for your reply! If I want to get original volume of a structure in the image, how do I proceed? The model can only take a registered image(after affine transformation).

**Reply by cookpa** [reply]
The original volume would be measured in its original space with `LabelGeometryMeasures`

**Reply by qwang0225** [reply]
Sorry, I'm kind of new to ANTs. Can you show me how to use it in this case? My original image does not have labels. I trained a segmentation model to segment registered images where I would have labels. But I want to get the original volume of the segmented structure. Thanks

**Reply by cookpa** [reply, marked as answer]
For labels in the fixed space, if you want label volume in the moving space, you can binarize the label, warp to the moving space, and then sum up the values. Something like:

```
unique_labels = fixed_labels.numpy().unique()
volumes = {}

for label in unique_labels:
    if label == 0:  # Skip background
        continue

    # Create a binary image for the current label
    label_image = ants.threshold_image(fixed_labels, low=label, high=label, inval=1, outval=0)

    # Warp the label image to the moving space using inverse transforms
    warped_label_image = ants.apply_transforms(fixed=moving_image, moving=label_image,      
        transformlist=transform['fwdtransforms'][0], whichtoinvert = [True])

    # Compute the volume
    voxel_size = moving_image.spacing  # Use the spacing of the moving image
    volume = warped_label_image.numpy().sum() * voxel_size[0] * voxel_size[1] * voxel_size[2]
    volumes[label] = volume
```

**Reply by qwang0225** [reply]
Thanks a lot!
